service/worker/dlq/workflow.go

```diff
@@ -0,0 +1,161 @@
+	WorkflowTypeDelete = "delete"
+
+	errorTypeInvalidWorkflowType = "dlq-invalid-workflow-type"
+	errorTypeUnavailable         = "history-service-unavailable"
+	errorTypeUnknown             = "unknown-error"
+	deleteTasksActivityName      = "dlq-delete-tasks-activity"
+
+	deleteTasksTimeout = 15 * time.Second
+)
+
+// Module provides a [workercommon.WorkerComponent] annotated with [workercommon.WorkerComponentTag] to the graph, given
+// a [HistoryServiceClient] dependency.
+var Module = workercommon.AnnotateWorkerComponentProvider(newComponent)
+
+func newComponent(client HistoryServiceClient) workercommon.WorkerComponent {
+	return &workerComponent{
+		client: client,
+	}
+}
+
+//revive:disable:import-shadowing this doesn't actually shadow imports because it's a method, not a function
+func (c *workerComponent) workflow(ctx workflow.Context, params WorkflowParams) error {
+	if params.WorkflowType == WorkflowTypeDelete {
+		ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
+			StartToCloseTimeout: deleteTasksTimeout,
+		})
+		future := workflow.ExecuteActivity(ctx, deleteTasksActivityName, params.DeleteParams)
+		return future.Get(ctx, nil)
+	}
+	return temporal.NewNonRetryableApplicationError(params.WorkflowType, errorTypeInvalidWorkflowType, nil)
+}
+
+func (c *workerComponent) deleteTasks(ctx context.Context, params DeleteParams) error {
+	req := &historyservice.DeleteDLQTasksRequest{
+		DlqKey: &commonspb.HistoryDLQKey{
+			TaskCategory:  int32(params.TaskCategory),
+			SourceCluster: params.SourceCluster,
+			TargetCluster: params.TargetCluster,
+		},
+		InclusiveMaxTaskMetadata: &commonspb.HistoryDLQTaskMetadata{
+			MessageId: params.MaxMessageID,
+		},
+	}
+	_, err := c.client.DeleteDLQTasks(ctx, req)
+	if err != nil {
+		var svcErr serviceerror.ServiceError
+		// Use this instead of serviceerror.ToStatus because that doesn't work for wrapped errors
+		if errors.As(err, &svcErr) {
+			if svcErr.Status().Code() == codes.Unavailable {
+				return temporal.NewApplicationErrorWithCause(
+					"delete tasks activity failed with unavailable error",
+					errorTypeUnavailable,
+					err,
+				)
+			}
+		}
+		// Consider all other errors as non-retryable
+		return temporal.NewNonRetryableApplicationError("error deleting DLQ tasks", errorTypeUnknown, err)
+	}
+	return err
+}
+
+func (c *workerComponent) RegisterWorkflow(registry sdkworker.Registry) {
+	registry.RegisterWorkflowWithOptions(c.workflow, workflow.RegisterOptions{
+		Name: WorkflowName,
+	})
+}
+
+func (c *workerComponent) DedicatedWorkflowWorkerOptions() *workercommon.DedicatedWorkerOptions {
+	// use default worker
+	return nil
+}
+
+func (c *workerComponent) RegisterActivities(registry sdkworker.Registry) {
+	registry.RegisterActivityWithOptions(c.deleteTasks, activity.RegisterOptions{
+		Name: deleteTasksActivityName,
+	})
+}
+
+func (c *workerComponent) DedicatedActivityWorkerOptions() *workercommon.DedicatedWorkerOptions {
+	// use default worker
```

Comment:
I forget what callerType is used by default worker, likely CallerTypeBackground (which is for active task processing). 

Idea is the actual process of merging/purging DLQ should use the lowest priority (CallerTypePreemptable). 
So either override the caller type in each activity implementation, or use a dedicated activity worker whose background activity context uses CallerTypePreemptable.

Reply:
Set the caller type to be preemptable.